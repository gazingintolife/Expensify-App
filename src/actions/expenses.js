import uuid from 'uuid';
import database from '../firebase/firebase';



export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

//startAddExpense is called by the component instead of addExpense as the former dispatches the later. 

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const { 
            description = '', 
            note = '',                              // destructuring data from expenseData
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

//REMOVE EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
    }
);

//EDIT EXPENSE
export const editExpense = (id, updates) => ({
        type: 'EDIT_EXPENSE',
        id,
        updates
    });
