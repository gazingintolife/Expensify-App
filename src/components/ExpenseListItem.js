import React from 'react';
import {NavLink} from 'react-router-dom';

// const ExpenseListItem = ({ id, description, amount, createdAt}) => (
    // <div>
        //  
        //  <NavLink to = {`/edit/${id}`} activeClassName='is-active' >Description: {description}</NavLink>
        //  <p>Amount: {amount}</p>
        //  <p>Created At: {createdAt}</p>
    // </div>
// );

const ExpenseListItem = (expenses) => (
    <div>
        <NavLink to = {`/edit/${expenses.id}`} activeClassName='is-active' >Description: {expenses.description}</NavLink>
        <p>Amount: {expenses.amount}</p>
        <p>Created At: {expenses.createdAt}</p>
    </div>
);

// this snippet of code below is same as code above as above we use desturcturing 

export default ExpenseListItem;