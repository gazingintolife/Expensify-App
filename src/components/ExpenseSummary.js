import React from 'react';
import numeral from 'numeral';
import {connect} from 'react-redux';
import SelectVisibleExpenses from '../selectors/expenses';
import getExpenseTotal from '../selectors/expensesTotal';

export const ExpensesSummary = ({count, total}) => {
    const formattedTotal = numeral(total/100).format('$0,0.00');
    const expenseWord = count === 1 ? 'expense' : 'expenses';
    return (
        <div>
            <p>Viewing {count} {expenseWord} totalling {formattedTotal}</p>
        </div>
    );
}

const mapStateToProps = (state) => {
    const visibleExpenses = SelectVisibleExpenses(state.expenses,state.filters)
    return {
        count: visibleExpenses.length,
        total: getExpenseTotal(visibleExpenses)
    };
}

export default connect(mapStateToProps)(ExpensesSummary);