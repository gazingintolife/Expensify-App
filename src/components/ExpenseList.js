import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
// selectExpenses is actually getVisibleExpenses but we default export it there

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expenses</p>
            ):
            (
                props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id} {...expense}/>;
                })
            )
        }
        
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses,state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);
