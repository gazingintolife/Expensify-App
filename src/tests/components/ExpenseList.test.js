import { ExpenseList } from '../../components/ExpenseList';
import React from 'react';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses';

test('should render expense list with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses = {expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render expense list empty message', () => {
    const wrapper = shallow(<ExpenseList expenses = {[]} />);
    expect(wrapper).toMatchSnapshot();
});