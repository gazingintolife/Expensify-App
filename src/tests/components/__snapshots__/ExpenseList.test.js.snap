// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list empty message 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`should render expense list with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id="0"
    key="0"
    note=""
  />
  <ExpenseListItem
    amount={10000}
    createdAt={-345600000}
    description="rent"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={1500}
    createdAt={345600000}
    description="credit"
    id="2"
    key="2"
    note=""
  />
</div>
`;
